modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/causalitydatanodes/CausalityDataNodesEngine.java

```diff
@@ -0,0 +1,717 @@
+    }
+
+    /**
+     * Gets data nodes of the zone using causality token.
+     *
+     * <p>Return data nodes or throw the exception:
+     * {@link IllegalArgumentException} if causalityToken or zoneId is not valid.
+     * {@link DistributionZoneNotFoundException} if the zone with the provided zoneId does not exist.
+     *
+     * @param causalityToken Causality token.
+     * @param zoneId Zone id.
+     * @return The data nodes for the zoneId.
+     */
+    public Set<String> dataNodes(long causalityToken, int zoneId) {
+        if (causalityToken < 1) {
+            throw new IllegalArgumentException("causalityToken must be greater then zero [causalityToken=" + causalityToken + '"');
+        }
+
+        if (zoneId < DEFAULT_ZONE_ID) {
+            throw new IllegalArgumentException("zoneId cannot be a negative number [zoneId=" + zoneId + '"');
+        }
+
+        if (!busyLock.enterBusy()) {
+            throw new IgniteException(NODE_STOPPING_ERR, new NodeStoppingException());
+        }
+
+        try {
+            ConcurrentSkipListMap<Long, ZoneConfiguration> versionedCfg = zonesVersionedCfg.get(zoneId);
+
+            // Get the latest configuration and configuration revision for a given causality token
+            Map.Entry<Long, ZoneConfiguration> zoneLastCfgEntry = versionedCfg.floorEntry(causalityToken);
+
+            if (zoneLastCfgEntry == null) {
+                // It means that the zone does not exist on a given causality token.
+                throw new DistributionZoneNotFoundException(zoneId);
+            }
+
+            long lastCfgRevision = zoneLastCfgEntry.getKey();
+
+            ZoneConfiguration zoneLastCfg = zoneLastCfgEntry.getValue();
+
+            String filter = zoneLastCfg.getFilter();
+
+            boolean isZoneRemoved = zoneLastCfg.getIsRemoved();
+
+            if (isZoneRemoved) {
+                // It means that the zone was removed on a given causality token.
+                throw new DistributionZoneNotFoundException(zoneId);
+            }
+
+            // Get revisions of the last scale up and scale down event which triggered immediate data nodes recalculation.
+            long lastScaleUpRevision = getRevisionsOfLastScaleUpEvent(causalityToken, zoneId);
+            long lastScaleDownRevision = getRevisionsOfLastScaleDownEvent(causalityToken, zoneId);
+
+            if (lastCfgRevision == versionedCfg.firstKey()
+                    && lastCfgRevision >= lastScaleUpRevision
+                    && lastCfgRevision >= lastScaleDownRevision
+            ) {
+                // It means that the zone was created but the data nodes value had not updated yet.
+                // So the data nodes value will be equals to the logical topology on the lastCfgRevision.
+
+                Entry topologyEntry = msManager.getLocally(zonesLogicalTopologyKey(), zoneLastCfgEntry.getKey());
+
+                if (topologyEntry.value() == null) {
```

Comment:
In general, it seems to be incorrect - we have working cluster and no nodes in logical topology. Please add TODO with IGNITE-20053 to process it, or even assertion, if it wouldn't cause the tests failure (most likely it shouldnt)

Reply:
I agree that topologyEntry must be not null here. I added assert.